### PR TITLE
[GlobalISel] replace right identity X * -1.0 with fneg(x)

### PR DESCRIPTION
**Pull request by nickleus27**
follow up patch to #78673

@Pierre-vh @jayfoad @arsenm Could you review when you have a chance.

### CODE REVIEW

File: llvm/test/CodeGen/X86/popcnt.ll [hunk 1]
```diff
@@ -43,6 +43,7 @@ define i8 @cnt8(i8 %x) nounwind readnone {
 ; X64-POPCNT-NEXT:    popcntl %eax, %eax
 ; X64-POPCNT-NEXT:    # kill: def $al killed $al killed $eax
 ; X64-POPCNT-NEXT:    retq
+
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/popcnt.ll:46]: The change seems unrelated, should better to revert.

**Review by jayfoad — CHANGES_REQUESTED** [review]
I don't think this is safe since fmul is an arithmetic operation that may quieten nans and flush denormals, whereas fneg is defined to just flip the sign bit.

See https://reviews.llvm.org/D109446 for the DAG equivalent.

**Comment by arsenm** [comment]
> I don't think this is safe since fmul is an arithmetic operation that may quieten nans and flush denormals, whereas fneg is defined to just flip the sign bit.
> 
> See https://reviews.llvm.org/D109446 for the DAG equivalent.

This is OK, we are allowed to drop canonicalizes for non-strictfp functions. The DAG is too strict and I've been meaning to remove those restrictions

**Comment by tschuett** [comment]
I believe that we should move to a visitOpcode model like the Dag combiner eventually. If this combine is registered before, `constant_fold_fp_binop`, then we get a `fneg`.  Otherwise, we may have gotten a constant for `fmul 1000,0 * -1`.

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 2]
```diff
@@ -493,6 +493,13 @@ def right_identity_one_fp: GICombineRule<
   (apply (GIReplaceReg $dst, $x))
 >;
 
+def right_identity_neg_one_fp: GICombineRule<
+  (defs root:$dst),
+  (match (G_FMUL $dst, $x, $y):$root,
+    [{ return Helper.matchConstantFPOp(${y}, -1.0); }]),
```
[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:499]: Unfortunately not, it's also been on my list but I'm not sure how to handle them well. TableGen doesn't have float literals. I can see a few options:

- Add float literals to tablegen
- Parse them from strings using APFloat
- Only allow hex/integer immediates and bitconvert to FP

What do you prefer?

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:499]: Not sure how much work adding float literals would be, but I assume that parsing them with APFloat would be an implementation detail of that.

**Comment by nickleus27** [comment]
> This is OK, we are allowed to drop canonicalizes for non-strictfp functions. The DAG is too strict and I've been meaning to remove those restrictions

Does this mean the matcher should be checking for FMafn flags or check that the function has UnsafeFPMath?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fpneg-one-fneg.mir [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:_(s32) = COPY $sgpr0
+    ; CHECK-NEXT: [[C:%[0-9]+]]:_(s32) = G_FCONSTANT float -5.000000e-01
+    ; CHECK-NEXT: [[FMUL:%[0-9]+]]:_(s32) = G_FMUL [[COPY]], [[C]]
+    ; CHECK-NEXT: $sgpr0 = COPY [[FMUL]](s32)
+    %0:_(s32) = COPY $sgpr0
+    %1:_(s32) = G_FCONSTANT float -0.5
+    %2:_(s32) = G_FMUL %0, %1
+    $sgpr0 = COPY %2(s32)
+
+...
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fpneg-one-fneg.mir:109]: Vector case wouldn't hurt

**Comment by arsenm** [comment]
> > This is OK, we are allowed to drop canonicalizes for non-strictfp functions. The DAG is too strict and I've been meaning to remove those restrictions
> 
> Does this mean the matcher should be checking for FMafn flags or check that the function has UnsafeFPMath?

Neither. It would be a question of checking the parent function's denormal mode (could maybe consider unsafe flags, but none are an exact match). I think we need to loosen the DAG code, and this is fine (for non-strictfp functions, in which case this would be G_STRICT_FMUL so this wouldn't fire anyway)

**Review by arsenm — APPROVED** [review]
InstCombine already does this, and the DAG should resume doing this

**Comment by Pierre-vh** [comment]
> I believe that we should move to a visitOpcode model like the Dag combiner eventually. If this combine is registered before, `constant_fold_fp_binop`, then we get a `fneg`. Otherwise, we may have gotten a constant for `fmul 1000,0 * -1`.

No, a `visitOpcode` model would be a regression compared to what we have now, IMHO.

I agree we need a way to tell the combiner which combine goes first. Currently I _think_ it's decided by the order in which the combines appear in the list, but I don't think it's actually enforced much (well, a bunch of tests check the match table as a whole so if its order changes, those tests will fail).

We can achieve ordering differently:

- Use a system like AddedComplexity, maybe with a bit more syntactic sugar, to tell which combine goes first.
- Allow a combine to explicitly mention some dependencies so it can't run unless something else has been checked first (probably overkill)
- etc.

Is it a concern/issue currently? Do we have bugs due to combine ordering?
If so, please create an issue. I don't mind looking into it and coming up with more detail proposals

**Comment by tschuett** [comment]
A miscompile is caused by ordering issues:
https://github.com/llvm/llvm-project/issues/78383
Nonetheless, I would expect UNDEF and catching misbeviour be executed/registered before optimizations.

**Comment by jayfoad** [comment]
> A miscompile is caused by ordering issues: #78383 Nonetheless, I would expect UNDEF and catching misbeviour be executed/registered before optimizations.

I think it's wrong to say that ordering issues _cause_ that problem. One of the combines is broken - that's what causes the problem. Ordering issues just happen to _hide_ the problem in some cases.

**Comment by tschuett** [comment]
Agreed that combine is broken, but detecting misbehaviour is executed/registered after the buggy optimisation.

**Comment by jayfoad** [comment]
No objection from me given @arsenm's comment about dropping canonicalizations. I thought this was already approved.

**Comment by nickleus27** [comment]
If everyone approves could someone go ahead and commit for me. I do not have write access. Thank you.

**Review by jayfoad — APPROVED** [review]

**Pull request merged**